File: September2022/Java/RemoveNthNodeFromEndOfList.java
// @saorav21994

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = new ListNode(-1);
        slow.next = head;
        if (head.next == null) return null;
        while (n > 1) {
            if (fast.next == null) { fast = null; break; }
            fast = fast.next;
            n -= 1;
        }
        // System.out.println(fast.val);
        int flag = 0;
        while (fast != null && fast.next != null) {
            flag = 1;
            fast = fast.next;
            slow = slow.next;
        }
        if (flag == 0) { return head.next; }
        else {
            slow.next = slow.next.next;
        }
        return head;
    }
}

File: September2022/C++/RemoveNthNodeFromEndOfList.cpp
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/
class Solution {
public:
    // TC: O(n)
    // SC: O(1)
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* p1 = head;
        ListNode* p2 = head;
        
        while (n > 0) {
            p1 = p1->next;
            if (p1 == NULL) {   // it means we have to delete the first node
                return head->next;
            }
            n--;
        }
        
        while (p1->next != NULL) {
            p1 = p1->next;
            p2 = p2->next;
        }
        
        p2->next = p2->next->next;
        return head;
    }
};
